### PR TITLE
Tl/lt 88456218 cleanup login.js

### PR DESCRIPTION
**Pull request by luketower**
[Pivotal Story](https://www.pivotaltracker.com/story/show/88456218)
- [x] Specs pass

Questions under investigation (Must be addressed before merging. Feel free to comment on PR with suggestions/answers):
- What is the best way to test this to make sure all functionality is still present?
- What other apps use this library that need to be modified?
- Is a version bump necessary for this change?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,8 +13,41 @@
...
+    "karma-phantomjs-launcher": "^0.1.0",
+    "karma-requirejs": "~0.2.0",
+    "karma-safari-launcher": "^0.1.1"
+  },
+  "scripts": {
+    "compile": "gulp build",
+    "build": "gulp dist",
+    "livereload": "gulp",
+    "watch": "gulp",
+    "watch:dev": "gulp",
```
Comment (michaelstalker) [inline package.json]: Do we need all these scripts? Three just run `gulp`.

File: src/error_handler.coffee [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+          formattedError = @_formatError key, value
+          messages += "<li>#{formattedError}</li>"
+          $form.find('.error input:first').focus()
+      else
+        messages += "An error has occurred."
+      @$box.append "<ul>#{messages}</ul>"
+      events.trigger('event/' + eventName, error)
+
+
+    _formatError: (key, value) ->
```
Comment (michaelstalker) [inline src/error_handler.coffee]: What are your thoughts about removing the underscore? It makes more sense for us to treat this as a public function on an `ErrorHandler` object than it did when this function lived on the `Login` object.

File: src/error_handler.coffee [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      switch key
+        when "base" then value
+        when "auth_key"
+          if value then value else ''
+        when "password_confirmation" then "Password confirmation #{value}"
+        else
+          formatted_key = new Formatter(key).sentenceCase()
+          if value then "#{formatted_key} #{value}" else ''
+
+    _clearErrors: ($div) ->
```
Comment (michaelstalker) [inline src/error_handler.coffee]: What are your thoughts about removing the underscore? It makes more sense for us to treat this as a public function on an `ErrorHandler` object than it did when this function lived on the `Login` object.

**Comment by michaelstalker** [comment]
> What other apps use this library that need to be modified?

AG, and maybe m.AG. @c0: Are there others?

> Is a version bump necessary for this change?

If you want other apps to include this change, yes. I don't think it would be bad to wait to include this change in other apps, though. Maybe we could wait until we change the public behavior of login.js.

File: src/error_handler.coffee [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  class ErrorHandler
+
+    constructor: (@error, @$box, @eventName) ->
+
+    generateErrors: () ->
+      @clearErrors()
+      messages = ''
+      if error?
+        $form = @$box.parent().find 'form'
+        $.each @error, (key, value) =>
```
Comment (c0) [inline src/error_handler.coffee]: Coffeescript supports looping out of the box:

`for key, value of @error`

File: src/formatter.coffee [hunk 5]
```diff
@@ -0,0 +1,12 @@
+define [], () ->
+  class Formatter
+
+    constructor: (@string) ->
+
+    sentenceCase: () ->
+      new_string = @string.replace('_', ' ')
```
Comment (c0) [inline src/formatter.coffee]: `#replace` only works on the first occurrence. Should this replace multiple `_`? e.g. `@string.split('_').join(' ')`

**Comment by Teapane** [comment]
- Looks like `Ag` and `mAg` both use this library.  
- We can create a symlink and check functionality on both Apps to verify. I can check `m.Ag` with these changes if you want to check `Ag`?  
- Both apps `bower.json` will need to be updated with the new version.  
- And all the things :star2: 
  - Michael Stalker asks, "Did you _really_ mean **all** the things?" :trollface: 
  - Tyler Long says, " :trollface: :trollface: **ALL** the things." 
  - Michael Stalker replies, "Dang. I guess you **did** mean all the things."

File: src/formatter.coffee (hunk 5 above)
Comment (Teapane) [inline src/formatter.coffee]: Good catch. :+1:

**Comment by Teapane** [comment]
Once this is merged, we will need to bump the version to `v0.2.29`. Will also require [PR 1](https://github.com/rentpath/ag/pull/2259) and [PR 2](https://github.com/rentpath/m.apartmentguide/pull/513) to be merged.

**Comment by luketower** [comment]
Please do not merge this yet. We're still solving some existing problems.

**Comment by luketower** [comment]
This PR is ready for more review and possible merge.

File: README.md [hunk 6]
```diff
@@ -1,14 +1,98 @@
-# login.js
+## Getting started for development.
```
Comment (michaelstalker) [inline README.md:1]: :+1: For all the documentation on the README

**Comment by michaelstalker** [comment]
I'll give this a :+1: .

The tests on the private functions were mostly already there, but sometime in the future I think it would be good to ask why we're testing these. They would be better off as public functions--either on the `Login` object or on another object--or untested.

**Pull request merged**